File: src/com/yiqiandewo/stringmatching/KMP.java
package com.yiqiandewo.stringmatching;

public class KMP {
    public static void main(String[] args) {
        String s = "BBC ABCDAB ABCDABCDABDE";
        String p = "ABCDABDF";
        System.out.println(kmpSearch(s, p));
    }

    public static boolean kmpSearch(String s, String p) {
        int[] next = getNext(p);
        int i = 0, j = 0;
        while (i < s.length() && j < p.length()) {
            while (j > 0 && s.charAt(i) != p.charAt(j)) {
                j = next[j -1];
            }

            if (s.charAt(i) == p.charAt(j)) {
                j++;
            }
            i++;
        }

        return j == p.length();
    }

    private static int[] getNext(String s) {
        int[] next = new int[s.length()];
        next[0] = 0;
        for (int i = 1, j = 0; i < next.length; i++) {

            while (j > 0 && s.charAt(i) != s.charAt(j)) {
                j = next[j - 1];
            }

            if (s.charAt(i) == s.charAt(j)) {
                j++;
            }

            next[i] = j;
        }

        return next;
    }


}

File: src/com/yiqiandewo/stringmatching/Kmp.ts
function getNext(str: string): number[] {
  const next: number[] = [];
  let k = -1;
  let j = 0;
  next[0] = -1;
  const len = str.length;
  while (j < len - 1) {
    if (k === -1 || str.charAt(j) === str.charAt(k)) {
      j++;
      k++;
      next[j] = k;
    } else {
      k = next[k];
    }
  }
  return next;
}

function kmpSearch(s: string, p: string): boolean {
  const next = getNext(p);
  const sLen = s.length;
  const pLen = p.length;
  let i = 0;
  let j = 0;
  while (i < sLen && j < pLen) {
    if (j === -1 || s.charAt(i) === p.charAt(j)) {
      i++;
      j++;
    } else {
      j = next[j];
    }
  }
  return j === pLen;
}

const res = kmpSearch("abcdababcdabd", "abcdabde");
console.log(res);
